crates/ide_completion/src/snippet.rs

```diff
@@ -0,0 +1,168 @@
+    }
+
+    pub fn is_item(&self) -> bool {
+        self.scope == SnippetScope::Item
+    }
+
+    pub fn is_expr(&self) -> bool {
+        self.scope == SnippetScope::Expr
+    }
+}
+
+impl PostfixSnippet {
+    pub fn new(
+        label: String,
+        snippet: &[String],
+        description: &[String],
+        requires: &[String],
+        scope: PostfixSnippetScope,
+    ) -> Option<Self> {
+        let (snippet, description) = validate_snippet(snippet, description, requires)?;
+        Some(PostfixSnippet {
+            scope,
+            label,
+            snippet,
+            description,
+            requires: requires.iter().cloned().collect(), // Box::into doesn't work as that has a Copy bound 😒
+        })
+    }
+
+    /// Returns None if the required items do not resolve.
+    pub(crate) fn imports(
+        &self,
+        ctx: &CompletionContext,
+        import_scope: &ImportScope,
+    ) -> Option<Vec<ImportEdit>> {
+        import_edits(ctx, import_scope, &self.requires)
+    }
+
+    pub fn snippet(&self, receiver: &str) -> String {
+        self.snippet.replace("$receiver", receiver)
+    }
+
+    pub fn is_item(&self) -> bool {
+        self.scope == PostfixSnippetScope::Type
+    }
+
+    pub fn is_expr(&self) -> bool {
+        self.scope == PostfixSnippetScope::Expr
+    }
+}
+
+fn import_edits(
+    ctx: &CompletionContext,
+    import_scope: &ImportScope,
+    requires: &[String],
+) -> Option<Vec<ImportEdit>> {
+    let resolve = |import| {
+        let path = ast::Path::parse(import).ok()?;
+        let item = match ctx.scope.speculative_resolve(&path)? {
+            hir::PathResolution::Macro(mac) => mac.into(),
+            hir::PathResolution::Def(def) => def.into(),
+            _ => return None,
+        };
+        let path = ctx.scope.module()?.find_use_path_prefixed(
+            ctx.db,
+            item,
+            ctx.config.insert_use.prefix_kind,
+        )?;
+        Some((path.len() > 1).then(|| ImportEdit {
+            import: LocatedImport::new(path.clone(), item, item, None),
+            scope: import_scope.clone(),
+        }))
+    };
+    let mut res = Vec::with_capacity(requires.len());
+    for import in requires {
+        match resolve(import) {
+            Some(first) => res.extend(first),
+            None => return None,
+        }
+    }
+    Some(res)
+}
+
+fn validate_snippet(
+    snippet: &[String],
+    description: &[String],
+    requires: &[String],
+) -> Option<(String, Option<String>)> {
+    // validate that these are indeed simple paths
+    // we can't save the paths unfortunately due to them not being Send+Sync
+    if requires.iter().any(|path| match ast::Path::parse(path) {
+        Ok(path) => path.segments().any(|seg| {
+            !matches!(seg.kind(), Some(ast::PathSegmentKind::Name(_)))
+                || seg.generic_arg_list().is_some()
+        }),
+        Err(_) => true,
+    }) {
+        return None;
+    }
```

Comment:
This does a simple validation once to filter out incorrect snippets, it would be great if we could store the `ast::Path` instead of discarding and reparsing on each completion for `speculative_resolution` after this step but `ast::Path` isn't `Send` so that doesn't work and I can't think of a way to make this work otherwise.